google-cloud-spanner/lib/google/cloud/spanner/admin/database.rb

```diff
@@ -0,0 +1,97 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "google-cloud-spanner"
+require "google/cloud/config"
+
+module Google
+  module Cloud
+    module Spanner
+      module Admin
+        module Database
+          ##
+          # Create a new client object for a DatabaseAdmin.
+          #
+          # This returns an instance of
+          # [Google::Cloud::Spanner::Admin::Database::V1::DatabaseAdmin::Client](https://googleapis.dev/ruby/google-cloud-spanner-admin-database-v1/latest/Google/Cloud/Spanner/Admin/Database/V1/DatabaseAdmin/Client.html)
+          # for version V1 of the API.
+          # However, you can specify specify a different API version by passing it in the
+          # `version` parameter. If the DatabaseAdmin service is
+          # supported by that API version, and the corresponding gem is available, the
+          # appropriate versioned client will be returned.
+          #
+          # ## About DatabaseAdmin
+          #
+          # Google Cloud Spanner Database Admin Service
+          #
+          # The Cloud Spanner Database Admin API can be used to create, drop, and
+          # list databases. It also enables updating the schema of pre-existing
+          # databases. It can be also used to create, delete and list backups for a
+          # database and to restore from an existing backup.
+          #
+          # @param version [::String, ::Symbol] The API version to connect to. Optional.
+          #   Defaults to `:v1`.
+          # @return [Admin::Database::V1::DatabaseAdmin::Client] A client object for the specified version.
+          #
+          def self.new
+            if configure.emulator_host
+              return Admin::Database::V1::DatabaseAdmin::Client.new do |config|
+                config.credentials = :this_channel_is_insecure
+                config.endpoint = configure.emulator_host
+              end
+            end
+
+            Admin::Database::V1::DatabaseAdmin::Client.new
```

Comment:
It looks to me as though the config object provided here by `Spanner::Admin::Database.configure` never makes it to the `Admin::Database::V1::DatabaseAdmin::Client` instance, which is quite different than how things work in [`Google::Cloud::Spanner.new`](), where fields such as `lib_name`, etc are actually used to configure the client that is returned. Nor does `Spanner::Admin::Database.new` provide a block parameter that is passed to the client's `new` method, as in [`Kms.key_management_service`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-kms/lib/google/cloud/kms.rb#L85). When using this method to create a client, no instance-scope configuration of that client would be possible here. (Note that `Kms.key_management_service` does not document the block param, which may be a potential problem in that library.)

Reply:
Thanks for the pointers @quartzmo. I now made the initialization look more like `Google::Cloud::Spanner.new`. However, I don't quite understand what to do with the block parameter now. Is this still needed, and if so, how do I use it? Sorry, I'm completely new to Ruby so if you have any code snippets or examples of how it can be done alongside other input parameters, that would be great.

Reply:
@skuruppu This is a question of style, but personally I think making this method consistent with `Google::Cloud::Spanner.new` is the right choice. Since introducing the micro-generator and the new clients that accept a block, the cleanest and easiest way to provide **unmodified** user configuration to them is simply to expose a block parameter that is passed to the client's `new` method, as in [`Kms.key_management_service`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-kms/lib/google/cloud/kms.rb#L85). But in this case, consistency with the long list of keyword args in `Google::Cloud::Spanner.new` is probably better for users who have experience using that method. It's a style call. We should probably leave this thread open for @dazuma to comment.

Reply:
Ah ok, that makes sense. I would love for @dazuma to chime in here as well then. I will work on the unit/accpetance tests in the meantime to make sure that what I've done works.